wire-ios-sync-engine/Source/Use cases/SearchUsersUseCase.swift

```diff
@@ -0,0 +1,109 @@
+//
+// Wire
+// Copyright (C) 2024 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireUtilities
+
+// sourcery: AutoMockable
+public protocol SearchUsersUseCaseProtocol {
+    func invoke(
+        query: String,
+        options: SearchOptions,
+        messageProtocol: MessageProtocol?
+    ) async throws -> SearchResult
+}
+
+public class SearchUsersUseCase: SearchUsersUseCaseProtocol {
+
+    // MARK: - Properties
+
+    private let context: NSManagedObjectContext
+    private let searchDirectory: SearchDirectory
```

Comment:
Does this need to be torn down in the deinit of the use case?

Reply:
I believe yes:
```
extension SearchDirectory: TearDownCapable {
    /// Tear down the SearchDirectory.
    ///
    /// NOTE: this must be called before releasing the instance
    public func tearDown() {
        // Evict all cached search users
        searchUsersCache?.removeAllObjects()

        // Reset search user observer center to remove unnecessarily observed search users
        contextProvider.viewContext.searchUserObserverCenter.reset()

        isTornDown = true
    }
}

```